### PR TITLE
[CALCITE-6300] Function MAP_VALUES/MAP_KEYS gives exception when mapVauleType and mapKeyType not equals map Biggest mapKeytype or mapValueType

### PR DESCRIPTION
**Pull request by caicancai**
https://issues.apache.org/jira/browse/CALCITE-6300

### CODE REVIEW

**Comment by caicancai** [comment]
I may need to add some comments, but the code structure is basically determined

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 1]
```diff
@@ -1149,6 +1149,56 @@ private static RelDataType mapReturnType(SqlOperatorBinding opBinding) {
...
+  private static RelDataType mapKeyReturnType(SqlOperatorBinding opBinding) {
+    List<RelDataType> operandType = new ArrayList<>();
+
+    operandType.add(opBinding.collectOperandTypes().get(0).getKeyType());
+    operandType.add(opBinding.collectOperandTypes().get(0).getValueType());
+    Pair<@Nullable RelDataType, @Nullable RelDataType> type =
+        getComponentTypes(
+            opBinding.getTypeFactory(), operandType);
+
+    requireNonNull(type, "inferred map type");
```
Comment (macroguo-ghy) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1162]: type is never null

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 2]
```diff
@@ -1149,6 +1149,56 @@ private static RelDataType mapReturnType(SqlOperatorBinding opBinding) {
...
+    requireNonNull(type.getKey(), "inferred map key type");
+    requireNonNull(type.getValue(), "inferred value type");
+
+    if (!type.getValue().getSqlTypeName().getName().equals("UNKNOWN")
```
Comment (macroguo-ghy) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1166]: Compare the `SqlTypeName` enum, not string

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 3]
```diff
@@ -1458,14 +1508,14 @@ private static RelDataType deriveTypeMapConcat(SqlOperatorBinding opBinding) {
   @LibraryOperator(libraries = {SPARK})
   public static final SqlFunction MAP_KEYS =
       SqlBasicFunction.create(SqlKind.MAP_KEYS,
-          ReturnTypes.TO_MAP_KEYS_NULLABLE,
```
Comment (macroguo-ghy) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1461]: remove `ReturnTypes.TO_MAP_KEYS_NULLABLE` if we don't use it any more.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 4]
```diff
@@ -1458,14 +1508,14 @@ private static RelDataType deriveTypeMapConcat(SqlOperatorBinding opBinding) {
...
+          SqlLibraryOperators::mapKeyReturnType,
           OperandTypes.MAP);
 
   /** The "MAP_VALUES(map)" function. */
   @LibraryOperator(libraries = {SPARK})
   public static final SqlFunction MAP_VALUES =
       SqlBasicFunction.create(SqlKind.MAP_VALUES,
-          ReturnTypes.TO_MAP_VALUES_NULLABLE,
```
Comment (macroguo-ghy) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1468]: remove it if we don't use it any more.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 5]
```diff
@@ -1149,6 +1149,56 @@ private static RelDataType mapReturnType(SqlOperatorBinding opBinding) {
...
+  private static RelDataType mapValueReturnType(SqlOperatorBinding opBinding) {
+    List<RelDataType> operandType = new ArrayList<>();
+
+    operandType.add(opBinding.collectOperandTypes().get(0).getKeyType());
+    operandType.add(opBinding.collectOperandTypes().get(0).getValueType());
+    Pair<@Nullable RelDataType, @Nullable RelDataType> type =
+        getComponentTypes(
+            opBinding.getTypeFactory(), operandType);
+
+    requireNonNull(type, "inferred map type");
```
Comment (macroguo-ghy) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1187]: type is never null

**Review by macroguo-ghy — COMMENTED** [review]
Before I read`adjustTypeForMapFunctionConstructor`, I want to know why we need this method? 
If `MAP` function throw an exception when component type not equal? Like `map('k1', cast(1 as tinyint), 'k2', cast(1 as bigint))`

**Comment by caicancai** [comment]
> Before I read`adjustTypeForMapFunctionConstructor`, I want to know why we need this method? If `MAP` function throw an exception when component type not equal? Like `map('k1', cast(1 as tinyint), 'k2', cast(1 as bigint))`

Thank you for such a quick review, I will write a plan to illustrate it

**Comment by caicancai** [comment]
Here you need to convert mapKeyType and mapValueType into the corresponding Biggest Type. According to the initial method, there is definitely no conversion of mapKeyType and mapValueType into the corresponding Biggest Type.
Why add an `adjustTypeForMapFunctionConstructor method`? Here I will give an example:

```java
map_values(map('foo', 1, 'bar', cast(1 as tinyint)))
```
test1
When performing node parsing, map('foo', 1, 'bar', cast(1 as tinyint)) will be regarded as a node, that is, `((SqlBasicCall) call.getOperandList().get(0)) `== map('foo', 1, 'bar', cast(1 as tinyint)),

I have to typecast mapKeyType and mapValueType via `((SqlBasicCall) call.getOperandList().get(0)).getOperandList()` to get what I want.

test2

I didn't find the corresponding method, so I wrote a new method

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 4 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1468]: I have deleted

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1468]: done, thanks

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 5 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: done, thanks

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 3 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1461]: done, thanks

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 2 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: done, thanks

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 1 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: done, thanks

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorUtil.java [hunk 6]
```diff
@@ -1381,6 +1399,54 @@ private static void adjustTypeForMultisetConstructor(
...
+   * if map's value targetType is double, this method would ensure that the value of the
+   * first map are cast to double.
+   *
+   * @param evenType the {@link RelDataType} to which the operands at even positions should be cast
+   * @param oddType the {@link RelDataType} to which the operands at odd positions should be cast
+   * @param sqlCallBinding the {@link SqlCallBinding} containing the operands to be adjusted
+   */
+  private static void adjustTypeForMapFunctionConstructor(
+      RelDataType evenType, RelDataType oddType, SqlCallBinding sqlCallBinding) {
+    SqlCall call = sqlCallBinding.getCall();
```
Comment (macroguo-ghy) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorUtil.java:1426]: Do you assume that `call` is a map constructor? I think the `call` can be a column , a complex expr (like `map_concat(map('foo', 1), map('bar', 2))`)...


If we have a table like
```sql
create table t (id int, map_column map<integer, integer>);
```

when we execute a query like
```sql
select map_keys(map_column) from t;
```

I think your code will throw an exeception. Can you add more tests for your PR?

**Review by macroguo-ghy — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorUtil.java (hunk 6 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorUtil.java:1426]: Good catch. Although my function currently only has map_values and map_keys use, but I think I should consider the situation you mentioned

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorUtil.java:1426]: @macroguo-ghy According to the official website, map_keys and map_values only allow map parameters.
捕获

I tried to execute this sql in spark-shell, but it reported an error.
捕获

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorUtil.java:1426]: But map_contact does have the bug you mentioned. Map cannot be converted to the corresponding biggest type.

**Comment by caicancai** [comment]
@macroguo-ghy I'm sorry, I made the wrong request. I thought I should resolve the conflict first

**Comment by caicancai** [comment]
@macroguo-ghy If you have time, you can help me take a look. I think it is reasonable to only recognize map. Thank you.

**Comment by macroguo-ghy** [comment]
> @macroguo-ghy If you have time, you can help me take a look. I think it is reasonable to only recognize map. Thank you.

Sorry, I'm very busy this week, I will review at the earliest this weekend. And you can ask others for help.

**Comment by caicancai** [comment]
> > @macroguo-ghy If you have time, you can help me take a look. I think it is reasonable to only recognize map. Thank you.
> 
> @caicancai Sorry, I'm very busy this week, I will review at the earliest this weekend. And you can ask others for help.

Small problem, this is open source, maybe I have more questions

**Comment by caicancai** [comment]
@mihaibudiu If you have time, can you review it for me? Thank you.

**Comment by caicancai** [comment]
@mihaibudiu Sorry for following up on this PR so late. If you have time, can you help me review this PR, thank you

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 7]
```diff
@@ -1163,6 +1163,52 @@ private static RelDataType mapReturnType(SqlOperatorBinding opBinding) {
...
+    operandType.add(opBinding.collectOperandTypes().get(0).getValueType());
+    Pair<@Nullable RelDataType, @Nullable RelDataType> type =
+        getComponentTypes(
+            opBinding.getTypeFactory(), operandType);
+
+    requireNonNull(type.left, () -> "type left of " + type.left);
+    requireNonNull(type.right, () -> "type right of " + type.right);
+    if (type.left.getSqlTypeName() != SqlTypeName.UNKNOWN
+        && type.right.getSqlTypeName() != SqlTypeName.UNKNOWN) {
+      SqlValidatorUtil.adjustTypeForMapFunctionConstructor(type, opBinding);
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1180]: Is this function allowed to mutate the operators?
Can you point to another example which does that?
I would not expect type inference to modify the program.

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1180]: I think what you mean is whether there are other functions like map_values and map_key functions. I will test it later to determine whether there are similar functions.

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1180]: map_contains_keys?
```bash
scala> val df = spark.sql("select map_contains_key(map(1, 'a', 2, 'b'), 1.0)")
val df: org.apache.spark.sql.DataFrame = [map_contains_key(map(1, a, 2, b), 1.0): boolean]

scala> df.show()
+--------------------------------------+
|map_contains_key(map(1, a, 2, b), 1.0)|
+--------------------------------------+
|                                  true|
+--------------------------------------+
```

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1180]: I think I should modify this jira description, it seems that not only these two functions have problems.
We can listen to other people's opinions, what do you think?
@tanclary

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1180]: I am asking whether the returnTypeInference function is allowed to change the opBinding.
Did you see other examples in the code where this happens?
Neither map_concat or map_entries do that.

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1180]: Sorry, I misunderstood what you meant. Of course there are other similar encodings
https://issues.apache.org/jira/browse/CALCITE-5976
https://issues.apache.org/jira/browse/CALCITE-5948

**Pull request closed without merging**